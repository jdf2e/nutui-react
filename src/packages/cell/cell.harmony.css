.nut-cell {
  position: relative;
  display: flex;
  flex-direction: row;
  width: 100%;
  line-height: 20px;
  padding: 13px 16px;
  background-color: #ffffff;
  border-radius: 6px;
  box-shadow: 0px 1px 7px 0px rgb(237, 238, 241);
  font-size: 14px;
  color: #1a1a1a;
  margin-bottom: 10px;
  box-sizing: border-box;
}
.nut-cell-group-item {
  border-radius: 0;
  box-shadow: 0 0 transparent;
  margin: 0;
}
.nut-cell-left {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  flex: 1;
}
.nut-cell-title {
  line-height: 20px;
}
.nut-cell-description {
  line-height: 20px;
}
.nut-cell-extra {
  line-height: 20px;
}
.nut-cell-description {
  font-size: 12px;
  color: #505259;
}
.nut-cell-extra {
  display: flex;
  flex-direction: row;
  justify-content: flex-end;
  align-items: center;
  flex: 1;
  flex-shrink: 0;
  min-width: 0;
  word-break: break-all;
  font-size: 14px;
  color: #505259;
}
.nut-cell-divider {
  display: flex;
  min-height: 1px;
  padding-left: 16px;
  padding-right: 16px;
}
.nut-cell-divider-inner {
  display: flex;
  height: 1px;
  width: 100%;
  border-top: 1px solid rgba(0, 0, 0, 0.06);
}
.nut-cell-divider-rtl {
  padding-left: 16px;
  padding-right: 16px;
}