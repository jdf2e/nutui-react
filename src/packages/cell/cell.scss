.nut-cell {
  position: relative;
  display: flex;
  flex-direction: row;
  width: 100%;
  line-height: $cell-line-height;
  padding: $cell-padding;
  background-color: $cell-background-color;
  border-radius: $cell-border-radius;
  box-shadow: $cell-box-shadow;
  font-size: $cell-title-font-size;
  color: $cell-title-color;
  margin-bottom: 10px;
  box-sizing: border-box;

  &-group-item {
    border-radius: 0;
    box-shadow: 0 0 transparent;
    margin: 0;
  }

  &-left {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    flex: 1;
  }

  &-description {
    font-size: $cell-description-font-size;
    color: $cell-description-color;
  }

  &-extra {
    display: flex;
    flex-direction: row;
    justify-content: flex-end;
    align-items: center;
    flex: 1;
    flex-shrink: 0;
    min-width: 0;
    word-break: break-all;
    font-size: $cell-extra-font-size;
    color: $cell-extra-color;
  }

  &-divider {
    display: flex;
    min-height: 1px;
    padding-left: $cell-divider-left;
    padding-right: $cell-divider-right;
    &-inner {
      display: flex;
      height: 1px;
      width: 100%;
      border-top: $cell-divider-border-bottom;
    }
    &-rtl {
      padding-left: $cell-divider-right;
      padding-right: $cell-divider-left;
    }
  }
}
