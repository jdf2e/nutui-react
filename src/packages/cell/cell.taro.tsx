import React, { FunctionComponent, ReactNode } from 'react'
import bem from '@/utils/bem'
import { BasicComponent, ComponentDefaults } from '@/utils/typings'

export interface CellProps extends BasicComponent {
  title: ReactNode
  description: ReactNode
  extra: ReactNode
  radius: string | number

  align: string
  className: string
  style: React.CSSProperties
  onClick: (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => void
}

const defaultProps = {
  ...ComponentDefaults,
  title: null,
  description: null,
  extra: null,
  radius: '6px',
  align: 'flex-start',
  className: '',
  style: {},
  onClick: (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => {},
} as CellProps

export const Cell: FunctionComponent<
  Partial<CellProps> & Omit<React.HTMLAttributes<HTMLDivElement>, 'title'>
> = (props) => {
  const {
    children,
    onClick,
    title,
    description,
    extra,
    radius,
    align,
    className,
    style,
    ...rest
  } = {
    ...defaultProps,
    ...props,
  }
  const b = bem('cell')
  const handleClick = (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    onClick(event)
  }

  const baseStyle = {
    ...style,
    borderRadius: Number.isNaN(Number(radius)) ? String(radius) : `${radius}px`,
    alignItems: align,
  }

  const styles =
    title || description
      ? {}
      : {
          flex: 1,
        }
  return (
    <div
      className={`${b({}, [className])} `}
      onClick={(event) => handleClick(event)}
      style={baseStyle}
      {...rest}
    >
      {children || (
        <>
          {title || description ? (
            <div className={`${b('title')}`}>
              {title ? <div className={b('maintitle')}>{title}</div> : null}
              {description ? (
                <div className={b('subtitle')}>{description}</div>
              ) : null}
            </div>
          ) : null}
          {extra ? (
            <div className={b('extra')} style={styles as React.CSSProperties}>
              {extra}
            </div>
          ) : null}
        </>
      )}
    </div>
  )
}

Cell.defaultProps = defaultProps
Cell.displayName = 'NutCell'
