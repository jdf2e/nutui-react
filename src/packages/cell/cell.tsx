import React, { FunctionComponent, ReactNode, useContext } from 'react'
import classNames from 'classnames'
import { BasicComponent, ComponentDefaults } from '@/utils/typings'
import CellGroup from '@/packages/cellgroup'
import CellGroupContext from '@/packages/cellgroup/context'

export interface CellProps extends BasicComponent {
  title: ReactNode
  description: ReactNode
  extra: ReactNode
  radius: string | number
  align: string
  onClick: (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => void
}

const defaultProps = {
  ...ComponentDefaults,
  title: null,
  description: null,
  extra: null,
  radius: '6px',
  align: 'flex-start',
  onClick: (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => {},
} as CellProps

const classPrefix = 'nut-cell'

export const Cell: FunctionComponent<
  Partial<CellProps> & Omit<React.HTMLAttributes<HTMLDivElement>, 'title'>
> & { Group: typeof CellGroup } = (props) => {
  const ctx = useContext(CellGroupContext)
  const {
    children,
    onClick,
    title,
    description,
    extra,
    radius,
    align,
    className,
    style,
    ...rest
  } = {
    ...defaultProps,
    ...props,
  }

  const handleClick = (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    onClick(event)
  }

  const baseStyle = {
    ...style,
    borderRadius: Number.isNaN(Number(radius)) ? String(radius) : `${radius}px`,
    alignItems: align,
  }

  const styles =
    title || description
      ? {}
      : {
          flex: 1,
        }
  return (
    <div
      className={classNames(classPrefix, className)}
      onClick={(event) => handleClick(event)}
      style={baseStyle}
      {...rest}
    >
      {children || (
        <>
          {title || description ? (
            <div className={`${classPrefix}-left`}>
              {title ? (
                <div className={`${classPrefix}-title`}>{title}</div>
              ) : null}
              {description ? (
                <div className={`${classPrefix}-description`}>
                  {description}
                </div>
              ) : null}
            </div>
          ) : null}
          {extra ? (
            <div
              className={`${classPrefix}-extra`}
              style={styles as React.CSSProperties}
            >
              {extra}
            </div>
          ) : null}
        </>
      )}
      {ctx?.divider ? <div className={`${classPrefix}-divider`} /> : null}
    </div>
  )
}

Cell.displayName = 'NutCell'
Cell.Group = CellGroup
