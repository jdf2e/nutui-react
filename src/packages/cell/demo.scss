.nutui-cell--clickable {
  &:active::before {
    opacity: 0.1;
  }
  cursor: pointer;
  &::before {
    position: absolute;
    top: 50%;
    left: 50%;
    width: 100%;
    height: 100%;
    background-color: $black;
    border: inherit;
    border-color: $black;
    border-radius: inherit;
    transform: translate(-50%, -50%);
    opacity: 0;
    content: ' ';
  }
}
.demo-extra {
  margin-right: '5px';
  flex: 1;
  min-width: 0;
}
