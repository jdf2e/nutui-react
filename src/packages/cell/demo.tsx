import React from 'react'
import { My } from '@nutui/icons-react'
import { useTranslate } from '../../sites/assets/locale'
import { Cell } from './cell'
import { CellGroup } from '../cellgroup/cellgroup'
import { Switch } from '../switch/switch'

interface T {
  basic: string
  desc: string
  desc1: string
  title: string
  title1: string
  title2: string
  title3: string
  title4: string
  title5: string
  title6: string
  title7: string
  title8: string
  title9: string
  link: string
  urlJump: string
  routerJump: string
  name: string
  image: string
  content: string
  customRight: string
  customLeftIcon: string
  displayIcon: string
}

const CellDemo = () => {
  const [translated] = useTranslate<T>({
    'zh-CN': {
      basic: '基本用法',
      desc: '描述文字',
      desc1: '使用 nut-cell-group 支持 title desc slots',
      title: '我是标题',
      title1: '副标题描述',
      title2: '直接使用插槽(slot)',
      title3: '点击测试',
      title4: '圆角设置 0',
      title5: '链接 | 分组用法',
      title6: '只展示 desc ，可通过 desc-text-align 调整内容位置',
      title7: '垂直居中',
      title8: '直接使用插槽(slot title)',
      title9: '尺寸设置 large',
      link: '链接',
      urlJump: 'URL 跳转',
      routerJump: '路由跳转 ’/‘ ',
      name: '姓名',
      image: '图片',
      content: '自定义内容',
      customRight: '自定义右侧箭头区域',
      customLeftIcon: '自定义左侧 Icon 区域',
      displayIcon: '展示图标',
    },
    'zh-TW': {
      basic: '基本用法',
      desc: '描述文字',
      desc1: '使用 nut-cell-group 支持 title desc slots',
      title: '我是標題',
      title1: '副標題描述',
      title2: '直接使用插槽(slot)',
      title3: '點擊測試',
      title4: '圓角設置 0',
      title5: '鏈接 | 分組用法',
      title6: '只展示 desc ，可通過 desc-text-align 調整內容位置',
      title7: '垂直居中',
      title8: '直接使用插槽(slot title)',
      title9: '尺寸設置 large',
      link: '鏈接',
      urlJump: 'URL 跳轉',
      routerJump: '路由跳轉 』/『 ',
      name: '姓名',
      image: '圖片',
      content: '自定義內容',
      customRight: '自定義右側箭頭區域',
      customLeftIcon: '自定義左側 Icon 區域',
      displayIcon: '展示圖標',
    },
    'en-US': {
      basic: 'Basic Usage',
      desc: 'Description',
      desc1: 'Usage nut-cell-group support title desc slots',
      title: 'Title',
      title1: 'Subtitle Description',
      title2: 'Use Slots',
      title3: 'Click Test',
      title4: 'Round Radius 0',
      title5: 'Link | CellGroup Usage',
      title6:
        'Only display desc , you can adjust the content position through desc-text-align',
      title7: 'Vertical Center',
      title8: 'Use Slots title',
      title9: 'Size setting large',
      link: 'Link',
      urlJump: 'URL Jump',
      routerJump: 'Router Jump ’/‘ ',
      name: 'Name',
      image: 'Image',
      content: 'Content',
      customRight: 'Customize the right arrow area',
      customLeftIcon: 'Customize the left Icon area',
      displayIcon: 'Display Icon',
    },
  })
  const testClick = (
    event: React.MouseEvent<HTMLDivElement, globalThis.MouseEvent>
  ) => {
    console.log('点击事件')
  }
  return (
    <>
      <div className="demo">
        <h2>{translated.basic}</h2>
        <Cell title={translated.title} desc={translated.desc} />
        <Cell
          title={translated.title}
          subTitle={translated.title1}
          desc={translated.desc}
        />
        <Cell
          title={translated.title3}
          onClick={(
            event: React.MouseEvent<HTMLDivElement, globalThis.MouseEvent>
          ) => testClick(event)}
        />
        <Cell title={translated.title4} roundRadius="0" />
        <h2>{translated.title9}</h2>
        <Cell size="large" title={translated.title} desc={translated.desc} />
        <Cell
          size="large"
          title={translated.title}
          subTitle={translated.title1}
          desc={translated.desc}
        />
        <h2>{translated.title2}</h2>
        <Cell>
          <div>{translated.content}</div>
        </Cell>
        <h2>{translated.title8}</h2>
        <Cell
          title={
            <span>
              Title <b style={{ color: 'red' }}>1</b>
            </span>
          }
          desc={translated.desc}
        />
        <CellGroup title={translated.title5} desc={translated.desc1}>
          <Cell title={translated.link} isLink />
          <Cell
            title={translated.urlJump}
            desc="https://jd.com"
            isLink
            url="https://jd.com"
          />
        </CellGroup>
        <CellGroup title={translated.customRight}>
          <Cell title="Switch" linkSlot={<Switch checked />} />
        </CellGroup>
        <CellGroup title={translated.customLeftIcon}>
          <Cell
            title={translated.image}
            icon={
              <img
                className="nut-icon"
                alt=""
                src="https://img11.360buyimg.com/imagetools/jfs/t1/137646/13/7132/1648/5f4c748bE43da8ddd/a3f06d51dcae7b60.png"
              />
            }
          />
        </CellGroup>
        <h2>{translated.displayIcon}</h2>
        <Cell
          title={translated.name}
          icon={<My />}
          desc={translated.desc}
          isLink
        />
        <h2>{translated.title6}</h2>
        <Cell descTextAlign="left" desc={translated.desc} />
        <h2>{translated.title7}</h2>
        <Cell
          center
          title={translated.title}
          subTitle={translated.title1}
          desc={translated.desc}
        />
      </div>
    </>
  )
}

export default CellDemo
