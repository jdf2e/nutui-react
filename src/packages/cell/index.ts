import { Cell } from './cell'

export type { CellProps } from './cell'
export default Cell
