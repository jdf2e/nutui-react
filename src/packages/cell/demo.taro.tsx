import React from 'react'
import Taro from '@tarojs/taro'
import { useTranslate } from '@/sites/assets/locale/taro'
import Header from '@/sites/components/header'
import '@/packages/cell/demo.scss'
import Demo1 from './demos/taro/demo1'
import Demo2 from './demos/taro/demo2'
import Demo3 from './demos/taro/demo3'
import Demo4 from './demos/taro/demo4'
import Demo5 from './demos/taro/demo5'
import Demo6 from './demos/taro/demo6'
import Demo7 from './demos/taro/demo7'

interface T {
  basic: string
  extra: string
  description: string
  title: string
  title1: string
  title2: string
  title3: string
  title4: string
  title5: string
  title6: string
  title7: string
  description1: string
  link: string
  urlJump: string
  content: string
  customRight: string
  clickEventToast: string
}

const CellDemo = () => {
  const [translated] = useTranslate<T>({
    'zh-CN': {
      basic: '基础用法',
      extra: '描述文字',
      description: '使用 nut-cell-group 支持 title description',
      title: '我是标题',
      title1: '我是描述',
      title2: '点击测试',
      title3: '圆角设置为0',
      title4: '链接 | 分组用法',
      title5: '垂直居中',
      title6: '自定义标题区域',
      title7: '分组用法',
      description1: '单元格之间不显示下边线',
      link: '链接',
      urlJump: 'URL 跳转',
      content: '自定义内容',
      customRight: '自定义右侧区域',
      clickEventToast: '点击事件',
    },
    'zh-TW': {
      basic: '基础用法',
      extra: '描述文字',
      description: '使用 nut-cell-group 支持 title description',
      title: '我是標題',
      title1: '我是描述',
      title2: '點擊測試',
      title3: '圓角設置為0',
      title4: '鏈接 | 分組用法',
      title5: '垂直居中',
      title6: '自定義標題區域',
      title7: '分組用法',
      description1: '單元格之間不顯示下邊線',
      link: '鏈接',
      urlJump: 'URL 跳轉',
      content: '自定義內容',
      customRight: '自定義右側區域',
      clickEventToast: '点击事件',
    },
    'en-US': {
      basic: 'Basic Usage',
      extra: 'extra',
      description: 'Usage nut-cell-group support title description',
      title: 'Title',
      title1: 'Description',
      title2: 'Click Test',
      title3: 'Round Radius 0',
      title4: 'Link | Cell.Group Usage',
      title5: 'Vertical Center',
      title6: 'Customize the title area',
      title7: 'Grouping usage',
      description1: 'The bottom edge is not displayed between cells',
      link: 'Link',
      urlJump: 'URL Jump',
      content: 'Customize Content',
      customRight: 'Customize the right area',
      clickEventToast: 'Click Test',
    },
  })

  return (
    <>
      <Header />
      <div className={`demo ${Taro.getEnv() === 'WEB' ? 'web' : ''}`}>
        <h2>{translated.basic}</h2>
        <Demo1 />
        <h2>{translated.content}</h2>
        <Demo2 />
        <h2>{translated.title6}</h2>
        <Demo3 />
        <h2>{translated.customRight}</h2>
        <Demo4 />
        <h2>{translated.title5}</h2>
        <Demo5 />
        <h2>{translated.title4}</h2>
        <Demo6 />
        <h2>{translated.title7}</h2>
        <Demo7 />
      </div>
    </>
  )
}

export default CellDemo
