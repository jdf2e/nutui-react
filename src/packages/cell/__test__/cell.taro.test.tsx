import TestUtils from '@tarojs/test-utils-react'
import '@testing-library/jest-dom'
import { Cell } from '../cell.taro'

const testUtils = new TestUtils()
const execFn = (desc: string, component: any, props: object, cb: any) => {
  it(desc, async () => {
    await testUtils.mount(component, {
      props,
    })
    cb()
    testUtils.unmout()
  })
}
describe('cell-taro', () => {
  execFn(
    'prop title extra description test',
    Cell,
    {
      title: '我是标题',
      description: '我是描述',
      extra: '描述文字',
    },
    () => {
      expect(
        testUtils.queries.querySelector('.nut-cell-title')?.innerHTML
      ).toBe('我是标题')
      expect(
        testUtils.queries.querySelector('.nut-cell-description')?.innerHTML
      ).toBe('我是描述')
      expect(
        testUtils.queries.querySelector('.nut-cell-extra')?.innerHTML
      ).toBe('描述文字')
      expect(testUtils.html()).toMatchSnapshot()
    }
  )
  execFn(
    'prop',
    Cell,
    {
      title: 'URL 跳转',
      extra: 'https://m.jd.com/',
    },
    () => {
      expect(
        testUtils.queries.querySelector('.nut-cell-extra')
      ).toBeInTheDocument()
      expect(testUtils.html()).toMatchSnapshot()
    }
  )
  execFn(
    'prop',
    Cell,
    {
      title: 'URL 跳转',
      extra: 'https://m.jd.com/',
    },
    () => {
      expect(
        testUtils.queries.querySelector('.nut-cell-extra')
      ).toBeInTheDocument()
      expect(testUtils.html()).toMatchSnapshot()
    }
  )
})
