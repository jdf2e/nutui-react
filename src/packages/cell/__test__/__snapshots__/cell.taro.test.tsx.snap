// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cell-taro prop 1`] = `
"[36m<body>[39m
  [36m<div />[39m
  [36m<div>[39m
    [36m<div[39m
      [33mclass[39m=[32m"nut-cell"[39m
      [33mstyle[39m=[32m"border-radius: 6px; align-items: flex-start;"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m"nut-cell-left"[39m
      [36m>[39m
        [36m<div[39m
          [33mclass[39m=[32m"nut-cell-title"[39m
        [36m>[39m
          [0mURL Ë∑≥ËΩ¨[0m
        [36m</div>[39m
      [36m</div>[39m
      [36m<div[39m
        [33mclass[39m=[32m"nut-cell-extra"[39m
      [36m>[39m
        [0mhttps://m.jd.com/[0m
      [36m</div>[39m
    [36m</div>[39m
  [36m</div>[39m
[36m</body>[39m"
`;

exports[`cell-taro prop 2`] = `
"[36m<body>[39m
  [36m<div />[39m
  [36m<div />[39m
  [36m<div>[39m
    [36m<div[39m
      [33mclass[39m=[32m"nut-cell"[39m
      [33mstyle[39m=[32m"border-radius: 6px; align-items: flex-start;"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m"nut-cell-left"[39m
      [36m>[39m
        [36m<div[39m
          [33mclass[39m=[32m"nut-cell-title"[39m
        [36m>[39m
          [0mURL Ë∑≥ËΩ¨[0m
        [36m</div>[39m
      [36m</div>[39m
      [36m<div[39m
        [33mclass[39m=[32m"nut-cell-extra"[39m
      [36m>[39m
        [0mhttps://m.jd.com/[0m
      [36m</div>[39m
    [36m</div>[39m
  [36m</div>[39m
[36m</body>[39m"
`;

exports[`cell-taro prop title extra description test 1`] = `
"[36m<body>[39m
  [36m<div>[39m
    [36m<div[39m
      [33mclass[39m=[32m"nut-cell"[39m
      [33mstyle[39m=[32m"border-radius: 6px; align-items: flex-start;"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m"nut-cell-left"[39m
      [36m>[39m
        [36m<div[39m
          [33mclass[39m=[32m"nut-cell-title"[39m
        [36m>[39m
          [0mÊàëÊòØÊ†áÈ¢ò[0m
        [36m</div>[39m
        [36m<div[39m
          [33mclass[39m=[32m"nut-cell-description"[39m
        [36m>[39m
          [0mÊàëÊòØÊèèËø∞[0m
        [36m</div>[39m
      [36m</div>[39m
      [36m<div[39m
        [33mclass[39m=[32m"nut-cell-extra"[39m
      [36m>[39m
        [0mÊèèËø∞ÊñáÂ≠ó[0m
      [36m</div>[39m
    [36m</div>[39m
  [36m</div>[39m
[36m</body>[39m"
`;
